api/clients/v2/accountant.go

```diff
@@ -5,115 +5,106 @@ import (
+	periodRecords     meterer.QuorumPeriodRecords
 	cumulativePayment *big.Int
 
 	// locks for concurrent access to period records and on-demand payment
 	periodRecordsLock sync.Mutex
 	onDemandLock      sync.Mutex
 }
 
-// PeriodRecord contains the index of the reservation period and the usage of the period
-type PeriodRecord struct {
-	// Index is start timestamp of the period in seconds; it is always a multiple of the reservation window
-	Index uint32
-	// Usage is the usage of the period in symbols
-	Usage uint64
-}
-
-func NewAccountant(accountID gethcommon.Address, reservation *core.ReservedPayment, onDemand *core.OnDemandPayment, reservationWindow uint64, pricePerSymbol uint64, minNumSymbols uint64, numBins uint32) *Accountant {
-	periodRecords := make([]PeriodRecord, max(numBins, uint32(meterer.MinNumBins)))
-	for i := range periodRecords {
-		periodRecords[i] = PeriodRecord{Index: uint32(i), Usage: 0}
-	}
-	a := Accountant{
-		accountID:         accountID,
-		reservation:       reservation,
-		onDemand:          onDemand,
-		reservationWindow: reservationWindow,
-		pricePerSymbol:    pricePerSymbol,
-		minNumSymbols:     minNumSymbols,
+// NewAccountant initializes an accountant with the given account ID. The accountant must call SetPaymentState to populate the state.
+func NewAccountant(accountID gethcommon.Address) *Accountant {
+	reservations := make(map[core.QuorumID]*core.ReservedPayment)
+	onDemand := &core.OnDemandPayment{
+		CumulativePayment: big.NewInt(0),
+	}
+	periodRecords := make(meterer.QuorumPeriodRecords)
+	return &Accountant{
+		accountID:    accountID,
+		reservations: reservations,
+		onDemand:     onDemand,
+		paymentVaultParams: &meterer.PaymentVaultParams{
+			QuorumPaymentConfigs:  make(map[uint8]*core.PaymentQuorumConfig),
+			QuorumProtocolConfigs: make(map[uint8]*core.PaymentQuorumProtocolConfig),
+			OnDemandQuorumNumbers: make([]uint8, 0),
+		},
 		periodRecords:     periodRecords,
 		cumulativePayment: big.NewInt(0),
 	}
-	// TODO: add a routine to refresh the on-chain state occasionally?
-	return &a
 }
 
-// reservationUsage attempts to use the reservation for the given request.
-func (a *Accountant) reservationUsage(
-	symbolUsage uint64,
-	quorumNumbers []uint8,
-	timestamp int64) error {
-	if err := meterer.ValidateQuorum(quorumNumbers, a.reservation.QuorumNumbers); err != nil {
+// ReservationUsage attempts to use the reservation for the requested quorums; if any quorum fails to use the reservation, the entire operation is rolled back.
+func (a *Accountant) reservationUsage(numSymbols uint64, quorumNumbers []core.QuorumID, timestamp int64) error {
+	// The two timestamps are the same for the accountant client for validating the reservation period; the second timestamp is the received at time for the server
+	if err := meterer.ValidateReservations(a.reservations, a.paymentVaultParams.QuorumProtocolConfigs, quorumNumbers, timestamp, time.Unix(0, timestamp)); err != nil {
```

Comment:
not following what this comment is saying. Also if you are using `time.Unix(0, timestamp)` then you have an assumption that the `timestamp int64` argument is in nanoseconds. So prob rename to `timestampNs`?

Reply:
```suggestion
func (a *Accountant) reservationUsage(numSymbols uint64, quorumNumbers []core.QuorumID, timestampNs int64) error {
	// The two timestamps are the same for the accountant client for validating the reservation period; the second timestamp is the received at time for the server
	if err := meterer.ValidateReservations(a.reservations, a.paymentVaultParams.QuorumProtocolConfigs, quorumNumbers, timestampNs, time.Unix(0, timestampNs)); err != nil {
```